### ISSUE TITLE
Best way to calculate loudness?

### ISSUE THREAD

**Issue by Cortexelus** [librosa/librosa#463]
What's the best way to calculate loudness per frame? And can we make it a function?

## TLDR;

Is it this?

``` python 
S = abs(FFT(y))**2 # power spectrogram
weighting = A_weighting # weighting in dB
weighting = 10**(weighting/10) # weighting for power spectrogram 
S *= weighting # perceptually weighted power spectrogram
S = melspectrogram(S) # perceptual pitch distances
loudness = np.mean(S) # taking mean is ok, because not in dB
loudness = logamplitude(loudness) # convert to dB
```

## Longer Explanation

Say we have a mel spectrogram
```python
y = load(mp3)
S = np.abs(FFT(y)) # magnitude spectrogram
S = S**2 # power spectrogram
S = melspectrogram(S) # convert frequencies to mel scale
```

We can weight the perceived loudness of frequencies differently this way:
```python
dB = logamplitude(S) 
perceptual_weighting = A_weighting * logamplitude(S)
```

But for loudness, we don't want the mean of dB values:
```
loudness = mean(perceptual_weighting)
```

because dB adds up like this
![image](https://cloud.githubusercontent.com/assets/2453861/20916298/075b7e0e-bb59-11e6-86a2-8a0145aa31af.png)

Can we use this equation to add up the dB frequencies to get loudness? It's more expensive because of extra squaring and division. 

OR

Can we weight the power spectrogram instead? Then taking the mean is fine, and we can calculate the dB from there: 
```python 
perceptual_weighting = A_weighting * power
loudness = logamplitude(mean(perceptual_weighting))
```
![image](https://cloud.githubusercontent.com/assets/2453861/20918269/2e50e7c8-bb64-11e6-8477-dffaf87573e4.png)

But A_weighting is a logarithmic weighting of dBs, so we need to convert it for power spectrograms

```python
weighting = 10**(A_weighting/10) # is this correct?
```
![image](https://cloud.githubusercontent.com/assets/2453861/20918282/3c97aae2-bb64-11e6-9273-6862fa4c987f.png)

And now we can calculate the mean just fine:
```python
loudness = np.mean(power * weighting)
loudness = logamplitude(loudness) # in db
```

This is great, because we also want to do weighting *before* we calculate the mel spectrogram, otherwise we lose frequency precision, which sometimes makes the subbass disappear, as in this example:

![image](https://cloud.githubusercontent.com/assets/2453861/20918095/4b52cef0-bb63-11e6-99d5-8b7fcb6542a5.png)
![image](https://cloud.githubusercontent.com/assets/2453861/20918118/6204cda6-bb63-11e6-8ef8-8c0008f7ae01.png)

**Comment by stefan-balke** [comment]
Hey CJ,

I think loudness is a very complicated thing.

Do you know these standards?

http://www.tcelectronic.com/de/loudness/broadcast-standards/

**Comment by Cortexelus** [comment]
Hi Stefan, 

Replicating other loudness standards would be useful. I'm okay with A-weighting. I'm looking to get validation that 
- The above is an appropriate and efficient way to arrive at loudness from a spectrogram
- `10**(weighting/10)` is the correct way to convert a dB loudness curve for a power spectrogram 
- This is correct and useful enough for a PR

**Comment by bmcfee** [comment]
> We can weight the perceived loudness of frequencies differently this way:

@Cortexelus do you know about [`librosa.perceptual_weighting`](https://librosa.github.io/librosa/generated/librosa.core.perceptual_weighting.html?highlight=perceptual#librosa.core.perceptual_weighting)?  It works like `logamplitude` but applies A-weighting to the specified frequency basis; I think it's probably what you're looking for here.

**Comment by Cortexelus** [comment]
@bmcfee Not quite. I'm looking for 1D loudness and a fast/precise method of its calculation. It seems it's better to apply perceptual weighting to the power spectrogram _before_ calculating the melspectrogram and the logamplitude (see reasoning above)

**Comment by bmcfee** [comment]
> to apply perceptual weighting to the power spectrogram before calculating the melspectrogram

Yes, `perceptual_weighting` can do that, if you invert the log scaling:

```python
>>> S = librosa.stft(y)**2
>>> log_S = librosa.perceptual_weighting(S**2, librosa.fft_frequencies())
>>> melspec = librosa.feature.melspectrogram(S=10.0**(log_S/10))
```
Issue #421 addresses inverting `logamplitude`, and a PR there would be welcome.

If you want frame-wise loudness, you can compute `log_S` as above and then sum over the rows: `loudness = log_S.sum(axis=0, keepdims=True)`.

**Comment by Cortexelus** [comment]
> If you want frame-wise loudness, you can compute log_S as above and then sum over the rows: loudness = log_S.sum(axis=0, keepdims=True).

I thought decibels don't sum like that. You need to square the rows, divide by the square of the reference, take the log10 of that, times 10. 
![image](https://cloud.githubusercontent.com/assets/2453861/20930080/3d9025ec-bb9a-11e6-8238-fa8cda87327b.png)

**Comment by keunwoochoi** [comment]
Is A-weighting good enough for you? Otherwise you can take a look at https://github.com/keunwoochoi/perceptual_weighting, which is a perceptual weighting code based on ISO226.
@bmcfee I once was thinking about PR this stuff. What do you think?

**Comment by bmcfee** [comment]
> You need to square the rows, divide by the square of the reference, take the log10 of that, times 10.

Okay, that's easy enough as well.  Since you have a constant factor of `p0**2` in the summation, you can factor it out of the log just like any other reference power.
```python
pow = np.abs(S)**2
p_mean = np.sum(pow, axis=0, keepdims=True)
p_ref = np.max(P)  # or whatever other reference power you want to use
loudness = librosa.power_to_db(P, ref=p_ref)
```
If you're on the 0.4 series, replace the last line by
```python
loudness = librosa.logamplitude(P, ref_power=p_ref)
```



> @bmcfee I once was thinking about PR this stuff. What do you think?

Not sure -- what would the PR add exactly?  Different weighting curves?

**Comment by keunwoochoi** [comment]
> Different weighting curves?
Yes, curves depends on frequency *and* loudness. It is tricky to match the physical energy and energy computed in digital, but we can assume the full-scale amplitude ~= 110 dB or so.

**Comment by lostanlen** [comment]
@keunwoochoi FYI we've changed the name `logamplitude` to `amplitude_to_db` in the latest release but a PR for ISO 226 is still welcome.

**Comment by DUOLabs333** [comment]
Sort of related, but is there a way to get the frequency-dB curves themselves?

**Comment by bmcfee** [comment]
Do you mean something like this? https://librosa.org/doc/latest/generated/librosa.multi_frequency_weighting.html

**Comment by DUOLabs333** [comment]
Yes, thank you.

**Comment by DUOLabs333** [comment]
@bmcfee Should the output of the function be interpreted as values to add to the amplitude of each frequency (like what is shown in Wikipedia), or something else?

**Comment by bmcfee** [comment]
> Should the output of the function be interpreted as values to add to the amplitude of each frequency (like what is shown in Wikipedia), or something else?

Not to the amplitude, but to the decibel values.  Probably a better way to use this is via the `perceptual_weighting` function (as an alternative to `amplitude_to_db`) with the `kind` value specified.  See: https://librosa.org/doc/latest/generated/librosa.perceptual_weighting.html

**Comment by DUOLabs333** [comment]
Yeah, sorry --- meant decibel values.

**Comment by bmcfee** [comment]
I don't know how we never closed this one out, but I think it's essentially addressed by #1134 ... six years ago
